CMake: option to disable swift in swift

etcwilde opened the pull request:
Adding `SWIFT_ENABLE_SWIFT_IN_SWIFT` option to enable or disable the parts of Swift that require a Swift compiler to build. This is meant for bootstrapping compilers on new platforms and is not guaranteed to result in a compiler that will pass the test suite.

This option is on by default so that folks won't forget. If the option is off, the resulting compiler does not include the Swift optimizer sources in SwiftCompilerSources nor does the resulting compiler have swift macro support.

DougGregor committed:
Build compiler and runtimes without Swift

Remove dependency on macros from compiler and stdlib build when
bootstrapping the compiler without Swift available.

etcwilde committed:
CMake: option to disable swift in swift

Adding `SWIFT_ENABLE_SWIFT_IN_SWIFT` option to enable or disable the
parts of Swift that require a Swift compiler to build. This is meant for
bootstrapping compilers on new platforms and is not guaranteed to result
in a compiler that will pass the test suite.

This option is on by default so that folks won't forget.
If the option is off, the resulting compiler does not include the Swift
optimizer sources in SwiftCompilerSources nor does the resulting
compiler have swift macro support.

etcwilde commented:
This will require the compiler and stdlib-side changes in https://github.com/swiftlang/swift/pull/77815
Without them, we have several build failures due to missing symbols: 
```
/sources/swift/lib/AST/ASTScope.cpp:162: error: undefined reference to 'swift_ASTGen_validateUnqualifiedLookup'                              
/sources/swift/lib/Sema/TypeChecker.cpp:771: error: undefined reference to 'swift_ASTGen_evaluatePoundIfCondition'                           
/sources/swift/include/swift/Basic/BasicBridgingImpl.h:149: error: undefined reference to 'swift_ASTGen_bridgedSwiftClosureCall_1'
```

compnerd commented on CMakeLists.txt:
```diff
@@ -939,41 +945,44 @@ set(SWIFT_MAIN_INCLUDE_DIR "${SWIFT_SOURCE_DIR}/include")
-    set(BOOTSTRAPPING_MODE "HOSTTOOLS")
+
+if(SWIFT_ENABLE_SWIFT_IN_SWIFT)
+  if("${SWIFT_NATIVE_SWIFT_TOOLS_PATH}" STREQUAL "")
```
Not your fault, but, I think that the idiomatic way would be:

```suggestion
  if(NOT SWIFT_NATIVE_SWIFT_TOOLS_PATH)
```

compnerd commented on localization/CMakeLists.txt:
```diff
@@ -1,35 +1,37 @@
-set(diagnostic_witness "${CMAKE_BINARY_DIR}/share/swift/diagnostics/generated")
+if(SWIFT_NATIVE_SWIFT_TOOLS_PATH)
```
Please don't do this. Do this at the higher level and just avoid entering the directory.

etcwilde committed:
Get stdlib building again

PR 79186 (https://github.com/swiftlang/swift/pull/79186) moved one of
the mandatory passes from the C++ implementation to the Swift
implementation resulting in a compiler that is unable to build the
standard library. The pass used to ensure that inaccessible control-flow
positions after an infinite loop was marked with `unreachable` in SIL.
Since the pass is no longer running, any function that returns a value
that also has an infinite loop internally must place a fatalError after
the infinite loop or it will fail to compile as the compiler will
determine that the function does not return from all control flow paths
even though some of the paths are unreachable.

etcwilde commented:
This patch is enough to bring up a stage-0 compiler without macro support using the C++ Swift driver without Swift on the platform.

Verified with on Ubuntu 24.04 without a pre-installed Swift:
```sh
./sources/swift/utils/build-script \
  --skip-test-cmark \
  --skip-test-lldb \
  --skip-test-swift \
  --skip-test-llbuild \
  --skip-test-swiftpm \
  --skip-test-swift-driver \
  --skip-test-xctest \
  --skip-test-foundation \
  --skip-test-libdispatch \
  --skip-test-playgroundsupport \
  --skip-test-indexstore-db \
  --skip-test-sourcekit-lsp \
  --skip-test-swiftdocc \
  --skip-test-wasm-stdlib \
  --skip-test-swiftformat \
  --swift-include-tests=0 \
  --llvm-include-tests=0 \
  --release \
  --skip-early-swiftsyntax \
  --skip-early-swift-driver \
  --skip-build-benchmarks \
  --build-runtime-with-host-compiler=0 \
  --build-swift-libexec=0 \
  --build-swift-remote-mirror=0 \
  '--swift-install-components=autolink-driver;compiler;clang-resource-dir-symlink;stdlib;sdk-overlay;toolchain-tools;license' \
  '--llvm-install-components=llvm-ar;llvm-ranlib;clang;clang-resource-headers;compiler-rt;clang-features-file' \
  '--extra-cmake-options=-DLLVM_TARGETS_TO_BUILD=X86;AArch64' \
  '--swift-cmake-options=-DSWIFT_ENABLE_SWIFT_IN_SWIFT:BOOL=NO -DSWIFT_INCLUDE_DOCS:BOOL=NO' \
  --build-subdir=stage0 \
  --install-destdir=/staging/stage0 \
  --install-llvm \
  --install-swift
```

The resulting compiler is able to build the standard library and a simple "hello world" program.

etcwilde commented:
@swift-ci please test

etcwilde commented on CMakeLists.txt:
```diff
@@ -939,41 +945,44 @@ set(SWIFT_MAIN_INCLUDE_DIR "${SWIFT_SOURCE_DIR}/include")
-    set(BOOTSTRAPPING_MODE "HOSTTOOLS")
+
+if(SWIFT_ENABLE_SWIFT_IN_SWIFT)
+  if(SWIFT_NATIVE_SWIFT_TOOLS_PATH)
```
Flipped condition.

```suggestion
  if(NOT SWIFT_NATIVE_SWIFT_TOOLS_PATH)
```

etcwilde commented:
@swift-ci please test

3405691582 commented:
FYI: when building Swift with a recent bootstrap compiler, I've noticed the following error:
```
Begin Error in Function: '$s11swiftASTGen0B7VisitorV22generateDeclAttributes_11allowStaticAC0eF6ResultVx_Sbt20_CompilerSwiftSyntax04WithfL0RzAH0m9ModifiersL0RzlF'                                                                              Error! Found a leak due to a consuming post-dominance failure!                  Value:   %255 = begin_borrow [lexical] %6 : ${ var BridgedDeclAttributes } // users: %309, %259                                                                 Post Dominating Failure Blocks:                                                  bb32                                                                           End Error in Function: '$s11swiftASTGen0B7VisitorV22generateDeclAttributes_11allowStaticAC0eF6ResultVx_Sbt20_CompilerSwiftSyntax04WithfL0RzAH0m9ModifiersL0RzlF'Found ownership error?! 
```
Bisecting leads to commit 008e10c2ec6b7a40928e292c2af89e1df8db10b2 and reverting this is a workaround for the problem. I haven't investigated or tested further than this, but a heads-up in case you run in to the same issue.

etcwilde commented:
yeah, I've seen that as well. Thanks for bisecting it.

DougGregor reviewed (approved):
Thank you for picking this up! Just the one question from me

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci please test Linux
